File: 377.combination-sum-iv.ts
/*
 * @lc app=leetcode id=377 lang=typescript
 *
 * [377] Combination Sum IV
 */

// @lc code=start
function combinationSum4(nums: number[], target: number): number {
  let dp = [1];

  for (let i = 1; i <= target; i++) {
    dp[i] = 0;
    for (let j = 0; j < nums.length; j++) {
      if (i >= nums[j]) {
        dp[i] += dp[i - nums[j]];
      }
    }
  }

  return dp[target];
}
// @lc code=end

File: 377.combination-sum-iv.rb
#
# @lc app=leetcode id=377 lang=ruby
#
# [377] Combination Sum IV
#

# @lc code=start
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}
# def combination_sum4(nums, target)
#   puts "nums: #{nums}, target: #{target}"
#   return 1 if target == 0
#
#   res = 0
#   (0...nums.size).each do |i|
#     res += combination_sum4(nums, target - nums[i]) if target >= nums[i]
#   end
#
#   res
# end
def combination_sum4(nums, target)
  dp = [1]

  (1..target).each do |i|
    dp[i] = 0
    nums.each { |num| dp[i] += dp[i - num] if i >= num }
  end

  puts dp.inspect
  dp[target]
end
# @lc code=end
